### ISSUE TITLE
[FEATURE] Add options to configure minimal score and single match score fo normalization processor for Hybrid search

### ISSUE THREAD

**Issue by martin-gaievski** [opensearch-project/neural-search#299]
### Is your feature request related to a problem?
There is no way of setting min score for a hit in the final result list from Hybrid query, as well as it's not possible to set a score that will be returned in case there is only one match from all sub-queries for a min-max normalization technique. That may affect relevance of the result. 

### What solution would you like?
For min score: set the minimal score a part of the configuration for normalization processor. Possible request can look like:

```
{
    "description": "Post processor for hybrid search",
    "phase_results_processors": [
        {
            "normalization-processor": {
                "normalization": {
                    "technique": "l2",
                    "parameters" : { 
                          "min_score" : 0.001 
                    }
                },
                "combination": {
                    "technique": "arithmetic_mean",
                    "parameters" : { 
                          "min_score" : 0.001 
                    }
                }
            }
        }
    ]
}
```

For a single match score: set the single match score as a parameter for normalization technique. For techniques that do not support such parameter is will be ignored. Possible request can look like:

```
{
    "description": "Post processor for hybrid search",
    "phase_results_processors": [
        {
            "normalization-processor": {
                "normalization": {
                    "technique": "min_max",
                    "parameters" : { 
                          "single_match_score" : 0.0 
                    }
                }
            }
        }
    ]
}
```

### What alternatives have you considered?
With current implementation additional post-processing is required.

### Do you have any additional context?
As part of the normalization scores are adjusted to be in the interval [0, .. 1.0]. That means matching doc with the minimal score will receive score 0.0. 
For a min-max normalization technique, if there is only one matching document there will be a single score, X. As per formula that leads to a "division by zero" case, as score is calculated as ```(X - min)/(max - min)```.

**Comment by heemin32** [comment]
@martin-gaievski, could you elaborate on why setting a minimum value is beneficial? You mentioned that *'As part of the normalization, scores are adjusted to be in the interval [0, .. 1.0]. That means a matching document with the minimal score will receive a score of 0.0.'* Would there be any issues if a document with the minimal score is assigned 0.0 after normalization?  

Regarding `single_match_score`, does it currently return Infinity? Wouldn't it be more intuitive to simply return 1 as the normalized score, rather than requiring the user to provide one?

**Comment by martin-gaievski** [comment]
> @martin-gaievski, could you elaborate on why setting a minimum value is beneficial? You mentioned that _'As part of the normalization, scores are adjusted to be in the interval [0, .. 1.0]. That means a matching document with the minimal score will receive a score of 0.0.'_ Would there be any issues if a document with the minimal score is assigned 0.0 after normalization?
Yes, it's an issue, as in most scenarios score 0.0 means this document is no match, so may drop a valid search hit.

> 
> Regarding `single_match_score`, does it currently return Infinity? Wouldn't it be more intuitive to simply return 1 as the normalized score, rather than requiring the user to provide one?
This currently makes sense for min/max normalization, and yes we do return 1.0 in case of a single matching doc. My idea was to keep default behavior same, but override if you provided a value for "single_match_score"

**Comment by jdomkline** [comment]
I love this feature request.  As a workaround, I'm using the following kludge that may be brittle.  I'm doing a HYB with a LEX and NEURAL sub queries

1) I wrap the LEX query in a script_score to lower the minscores down to 0.
              "script_score": {
                "query": {..},
                "script": {
                  "source": "if (_score >= 10) { return _score; } else { return 0; }"
                }
2) Then in the Search request I apply the minscore with the min in the Neural range, e.g.  "min_score" : 0.002.

This trims off both the desired LEX and Neural results.  Is it, however, too much of a hack?  I plan to recheck it with each release (I'm on AWS OSS v2.17)

**Comment by martin-gaievski** [comment]
@jdomkline are you planning to cut matching documents based on combined or individual sub-query (e.g. only neural query) score? if it's combined then `min_score` should work for you, just one problem I can think of is the exact threshold, it will be hard to predict for different queries. Or you rather need a min score for each sub-query?

**Comment by Gfr1992** [comment]
Hello it seems to me that the `min_score` is not applied to the final score computed by the pipeline, but to the individual score of each document extracted by the different clause of the hybrid search.

Below the pipeline and the hybrid search.


```
PUT /_search/pipeline/hybrid_semantic
{
  "description": "Post processor for hybrid search",
  "phase_results_processors": [
    {
      "normalization-processor": {
        "normalization": {
          "technique": "min_max"
        },
        "combination": {
          "technique": "arithmetic_mean",
          "parameters": {
            "weights": [
              0.5,
              0.5
            ]
          }
        }
      }
    }
  ]
}
```

```
POST index_name/_search?search_pipeline= hybrid_semantic
{
  "size": 1000,
  "_source": [
    "item_id",
    "item_description"],
  "query": {
    "bool": {
      "should": [
        {
          "knn": {
            "embedding": {
              "vector": [...],
              "k": 1000
            }
          }
        },
        {
          "match": {
            "item_description": {
              "query": "Description here",
              "analyzer": "italian",
              "fuzziness": 1,
              "operator": "and"
            }
          }
        }
      ]
    }
  },
  "min_score": 1.6
}
```

It seems to me that `min_score = 1.6`  is applied just to the results of the knn, since the combined score is between 0 an 1 and I get results with this filter. In addition, lowering the `min_score` to 1.5 I have more results coming by knn.
The index uses faiss for knn with inner product.

Can you help me?

**Comment by martin-gaievski** [comment]
@Gfr1992 you are correct, min_score behaves like this. This feature feature does not address your use case the way you described, while it's in design phase the way we're planning to implement it - user will be able to set a min score for each sub-query as part of the processor configuration. 

We can do the simple "cut off" logic for the final "after normalization" scores, that can be done with something like response processor. Can you please open a new feature request and our team will review and prioritize it.

**Comment by martin-gaievski** [comment]
I see folks are often using term "minimal score" for two different things, one is a `hybrid_query` specific term and defines the lower bound for the minimum score in min/max normalization technique, and second refers to a standard parameter for search queries (see `min_score` at this page https://opensearch.org/docs/latest/api-reference/search/#request-body).

This feature taking care of the first option, so I created new issue to address the second option https://github.com/opensearch-project/neural-search/issues/1164.

I suggest to call the first option a lower bound to avoid future confusion.

**Comment by minalsha** [comment]
@Gfr1992  could you please +1 on https://github.com/opensearch-project/neural-search/issues/1164 for us to help with prioritization of the feature request? Thanks

**Comment by jdomkline** [comment]
> [@jdomkline](https://github.com/jdomkline) are you planning to cut matching documents based on combined or individual sub-query (e.g. only neural query) score? if it's combined then `min_score` should work for you, just one problem I can think of is the exact threshold, it will be hard to predict for different queries. Or you rather need a min score for each sub-query?

@martin-gaievski I did end up using the script_score explicitly around both sub queries.  This allowed much better control.  I found some decent cutoff values.  This was not really about topside relevance, it is just about trimming the long tail of not-very-relevant docs.

**Comment by jdomkline** [comment]
I can run something crazy with the "clip" like below and get the same results (353) as when I use "ignore".  I'm not sure what was intended here... any numbers I try always the same result count.  The first query is a simple BM25 with a max score of 27.  I should ger zero results with this right?

        "normalization-processor": {
          "normalization": {
            "technique": "min_max",
            "parameters": {
              "lower_bounds": [
                {
                  "mode": "clip",
                  "min_score": 1000.0
                },
                {
                  "mode": "clip",
                  "min_score": 1000.00
                }
              ]
            }
          },
          "combination": {
            "technique": "arithmetic_mean",
            "parameters": {
              "weights": [
                0.3,
                0.7
              ]
            }
          }
        }

**Comment by jdomkline** [comment]
From the documentation
- clip: Replaces scores below the lower bound with min_score. Formula: min_max_score = **if (score < lowerBoundScore) then 0.0 else** (score - lowerBoundScore) / (maxScore - lowerBoundScore).    https://docs.opensearch.org/latest/search-plugins/search-pipelines/normalization-processor/#request-body-fields

**Comment by martin-gaievski** [comment]
> I can run something crazy with the "clip" like below and get the same results (353) as when I use "ignore". I'm not sure what was intended here... any numbers I try always the same result count. The first query is a simple BM25 with a max score of 27. I should get zero results with this right? I just see tiny changes in the scores.
> 
> ```
>     "normalization-processor": {
>       "normalization": {
>         "technique": "min_max",
>         "parameters": {
>           "lower_bounds": [
>             {
>               "mode": "clip",
>               "min_score": 1000.0
>             },
>             {
>               "mode": "clip",
>               "min_score": 1000.00
>             }
>           ]
>         }
>       },
>       "combination": {
>         "technique": "arithmetic_mean",
>         "parameters": {
>           "weights": [
>             0.3,
>             0.7
>           ]
>         }
>       }
>     }
> ```

"Clip" mode in the current implementation does not filter out documents based on the `min_score`. Instead, it clips the scores below the `min_score` to zero, which affects the normalized score but does not impact the result count. This behavior is by design, as the normalization technique focuses on adjusting the score values rather than filtering the documents.

To achieve the desired behavior of filtering out documents based on a score threshold, you may need to combine the normalization with an additional filter query or use a different approach, such as using a range filter in your query.
